Lib/zipimport.py

```diff
@@ -0,0 +1,640 @@
+
+This module exports three objects:
+- zipimporter: a class; its constructor takes a path to a Zip archive.
+- ZipImportError: exception raised by zipimporter objects. It's a
+  subclass of ImportError, so it can be caught as ImportError, too.
+- _zip_directory_cache: a dict, mapping archive paths to zip directory
+  info dicts, as used in zipimporter._files.
+
+It is usually not needed to use the zipimport module explicitly; it is
+used by the builtin import mechanism for sys.path items that are paths
+to Zip archives.
+'''
+
+#from importlib import _bootstrap_external
+#from importlib import _bootstrap  # for _verbose_message
+import _frozen_importlib_external as _bootstrap_external
+from _frozen_importlib_external import _unpack_uint16, _unpack_uint32
+import _frozen_importlib as _bootstrap  # for _verbose_message
+import _imp  # for check_hash_based_pycs
+import _io  # for open
+import marshal  # for loads
+import sys  # for modules
+import time  # for mktime
+
+__all__ = ['ZipImportError', 'zipimporter']
+
+
+path_sep = _bootstrap_external.path_sep
+alt_path_sep = _bootstrap_external.path_separators[1:]
+
+
+class ZipImportError(ImportError):
+    pass
+
+# _read_directory() cache
+_zip_directory_cache = {}
+
+_module_type = type(sys)
+
+
+class zipimporter:
+    '''zipimporter(archivepath) -> zipimporter object
+
+    Create a new zipimporter instance. 'archivepath' must be a path to
+    a zipfile, or to a specific path inside a zipfile. For example, it can be
+    '/tmp/myimport.zip', or '/tmp/myimport.zip/mydirectory', if mydirectory is a
+    valid directory inside the archive.
+
+    'ZipImportError is raised if 'archivepath' doesn't point to a valid Zip
+    archive.
+
+    The 'archive' attribute of zipimporter objects contains the name of the
+    zipfile targeted.
+    '''
+
+    # Split the "subdirectory" from the Zip archive path, lookup a matching
+    # entry in sys.path_importer_cache, fetch the file directory from there
+    # if found, or else read it from the archive.
+    def __init__(self, path):
+        if not isinstance(path, str):
+            import os
+            path = os.fsdecode(path)
+        if not path:
+            raise ZipImportError('archive path is empty', path=path)
+        if alt_path_sep:
+            path = path.replace(alt_path_sep, path_sep)
+
+        prefix = []
+        while True:
+            try:
+                st = _bootstrap_external._path_stat(path)
+            except (OSError, ValueError):
+                # On Windows a ValueError is raised for too long paths.
+                # Back up one path element.
+                dirname, basename = _bootstrap_external._path_split(path)
+                if dirname == path:
+                    raise ZipImportError('not a Zip file', path=path)
+                path = dirname
+                prefix.append(basename)
+            else:
+                # it exists
+                if (st.st_mode & 0o170000) != 0o100000:  # stat.S_ISREG
```

Comment:
Thinking about this a bit more - I'm totally cool with your eventual goal.  Can you put this in the docstring or file comment?  If this is a constraint, let's make it explicit (even better, through some tests), so that the constraint isn't accidentally broken later.